include/odp/api/spec/comp.h

```diff
@@ -0,0 +1,873 @@
+ *
+ * In case of ODP_COMP_ERR_OUT_OF_SPACE error, application should
+ * keep on calling odp_comp_op() API with more output buffer until call returns
+ * with success or any other error code.
+ *
+ * @param         pkt_in   Input Packet
+ * @param         pkt_out  Output Packet
+ * @param         param    Operation parameter
+ * @param[out] result   Result parameter, containing after operation result.
+ *
+ * @return 0 on success
+ * @retval <0 on failure
+ */
+int odp_comp_op(const odp_packet_t pkt_in,
+		const odp_packet_t pkt_out,
+		const odp_comp_packet_op_param_t *param,
+		odp_comp_packet_op_result_t *result);
+
+/**
+ * Comp API to process a packet asynchronously
+ *
+ * If session is created in ODP_COMP_ASYNC mode, result will be queued
+ * to completion queue. Application should monitor ODP_EVENT_PACKET
+ * with subtype ODP_EVENT_PACKET_COMP on queue.
+ *
+ * If session is created in ODP_COMP_SYNC mode, behavior is undefined.
+ *
+ * For compression + hash, call returns with hash appended at the end of
+ * packet which has last field set in corresponding param.
+ *
+ * User should compute processed data len = output_data_range.len - digest_len,
+ * where digest_len queried through odp_comp_hash_alg_capability().
+ *
+ * For every input packet, user should pass corresponding valid output
+ * packet.
+ *
+ * If API returns with success, application should monitor for
+ * ODP_EVENT_PACKET with subtype set to ODP_EVENT_PACKET_COMP on
+ * completion queue provided during session create.
+ * Once event is received, application should call odp_comp_result() to
+ * retrieve result corresponding to operation done on packet.
+ *
+ * For every successful or partially processed packet, api update result param
+ * with out_data_range indicating length of the data written in output packet
+ * and error code indicating operation status.
+ *
+ * A packet is said to be partially processed if it runs out of buffer during
+ * processing and need more buffer from user to continue. Under such condition,
+ * API fails with error code code in result param set to
+ * ODP_COMP_ERR_OUT_OF_SPACE.
+ *
+ * In case of ODP_COMP_ERR_OUT_OF_SPACE error, application should
+ * keep on calling odp_comp_op_enq() API with more output buffer until call
+ * returns with success or any other error code.
+ *
+ * Please note it is always recommended that application using async mode,
+ * provide sufficiently large buffer size to avoid ODP_COMP_ERR_OUT_OF_SPACE.
+ * Else it is recommended that application set required user context on packet
+ * before calling each odp_comp_op_enq() to correctly retrieve required
+ * information associated with packet.
+ *
+ * @param         pkt_in   Input Packet
+ * @param         pkt_out  Output Packet
+ * @param         param    Operation parameter
+ *
+ * @return 0 on success
+ * @retval <0 on failure
+ */
+int odp_comp_op_enq(const odp_packet_t pkt_in,
+		    const odp_packet_t pkt_out,
+		    const odp_comp_packet_op_param_t *param);
+
```

Comment:
First, async call can reuse sync call spec by saying: "... processes packets otherwise identically to odp_comp_op(), but outputs resulting packets as ODP_EVENT_PACKET events ...". Then the remaining text can concentrate on what is unique on async vs sync operation.

I think OUT_OF_SPACE spec might need re-thinking:
1) user sees a partial packet (error on a received packet)
2) implementation waits that user calls op_enq() again
3) what input packet should be given on the new call? I think the previous operation already consumed the input packet which contains the input data for operation. Did implementation store it?
4) how long implementation waits for new output packet to arrive from user, what if user don't send new output packets.

Async spec (and also sync) spec would be more robust if implementation would allocate output packets.